incubator/hnc/internal/forest/namespaceobjects.go

```diff
@@ -5,76 +5,77 @@ import (
 	}
-	ns.originalObjects[gvk][name] = obj
+	ns.sourceObjects[gvk][name] = obj
 }
 
-// GetOriginalObject gets an original object by name. It returns nil, if the object doesn't exist.
-func (ns *Namespace) GetOriginalObject(gvk schema.GroupVersionKind, nm string) *unstructured.Unstructured {
-	return ns.originalObjects[gvk][nm]
+// GetSourceObject gets a source object by name. If it doesn't exist, return nil.
+func (ns *Namespace) GetSourceObject(gvk schema.GroupVersionKind, nm string) *unstructured.Unstructured {
+	return ns.sourceObjects[gvk][nm]
 }
 
-// HasOriginalObject returns if the namespace has an original object.
-func (ns *Namespace) HasOriginalObject(gvk schema.GroupVersionKind, oo string) bool {
-	return ns.GetOriginalObject(gvk, oo) != nil
+// HasSourceObject returns if the namespace has a source object.
+func (ns *Namespace) HasSourceObject(gvk schema.GroupVersionKind, oo string) bool {
+	return ns.GetSourceObject(gvk, oo) != nil
 }
 
-// DeleteOriginalObject deletes an original object by name.
-func (ns *Namespace) DeleteOriginalObject(gvk schema.GroupVersionKind, nm string) {
-	delete(ns.originalObjects[gvk], nm)
+// DeleteSourceObject deletes a source object by name.
+func (ns *Namespace) DeleteSourceObject(gvk schema.GroupVersionKind, nm string) {
+	delete(ns.sourceObjects[gvk], nm)
 	// Garbage collection
-	if len(ns.originalObjects[gvk]) == 0 {
-		delete(ns.originalObjects, gvk)
+	if len(ns.sourceObjects[gvk]) == 0 {
+		delete(ns.sourceObjects, gvk)
 	}
 }
 
-// GetOriginalObjects returns all original objects in the namespace.
-func (ns *Namespace) GetOriginalObjects(gvk schema.GroupVersionKind) []*unstructured.Unstructured {
+// GetSourceObjects returns all source objects in the namespace.
+func (ns *Namespace) GetSourceObjects(gvk schema.GroupVersionKind) []*unstructured.Unstructured {
 	o := []*unstructured.Unstructured{}
-	for _, obj := range ns.originalObjects[gvk] {
+	for _, obj := range ns.sourceObjects[gvk] {
 		o = append(o, obj)
 	}
 	return o
 }
 
-// GetNumOriginalObjects returns the total number of original objects of a specific GVK in the namespace.
-func (ns *Namespace) GetNumOriginalObjects(gvk schema.GroupVersionKind) int {
-	return len(ns.originalObjects[gvk])
+// GetNumSourceObjects returns the total number of source objects of a specific
+// GVK in the namespace.
+func (ns *Namespace) GetNumSourceObjects(gvk schema.GroupVersionKind) int {
+	return len(ns.sourceObjects[gvk])
 }
 
-// GetPropagatedObjects returns all original copies in the ancestors.
-func (ns *Namespace) GetPropagatedObjects(gvk schema.GroupVersionKind) []*unstructured.Unstructured {
+// GetAncestorSourceObjects returns all source objects in the ancestors.
+func (ns *Namespace) GetAncestorSourceObjects(gvk schema.GroupVersionKind) []*unstructured.Unstructured {
 	o := []*unstructured.Unstructured{}
 	ans := ns.AncestryNames()
 	for _, n := range ans {
 		// Exclude the original objects in this namespace
 		if n == ns.name {
 			continue
 		}
```

Comment:
How about we remove lines 58-61, get rid of `GetAncestorAndLocalSourceObjects`, and then the (one current) caller to this function just says `ns.GetParent().GetAncestorSourceObjects()` instead of `ns.GetAncestorSourceObjects()`? You could change this function to return nil if `ns` is nil.

This would simplify the implementation and reduce the confusion between having _three_ similar functions instead of just two. Maybe at that point, you could have `GetLocalSourceObjects` and `GetAncestorSourceObjects`.

Also, [according to the Intro to Algorithms book](https://stackoverflow.com/a/3078231/158143), a node considered to be one of its own ancestors. So this is a defensible choice and also matches the behaviour of `ns.AncestryNames()` (maybe we should change that to `AncestorNames`).

Reply:
sgtm!